### PR TITLE
Enable konnectivity

### PR DESCRIPTION
**Pull request by sjenning**
@csrwng @relyt0925

### CODE REVIEW

File: control-plane-operator/controllers/hostedcontrolplane/pki/konnectivity.go [hunk 1]
```diff
@@ -32,6 +32,10 @@ func (p *PKIParams) ReconcileKonnectivityClusterSecret(secret, ca *corev1.Secret
 	return p.reconcileSignedCertWithAddresses(secret, ca, "konnectivity-server", "kubernetes", X509SignerUsage, X509UsageClientServerAuth, dnsNames, ips)
 }
 
+func (p *PKIParams) ReconcileKonnectivityClientSecret(secret, ca *corev1.Secret) error {
+	return p.reconcileSignedCertWithAddresses(secret, ca, "konnectivity-client", "kubernetes", X509SignerUsage, X509UsageClientServerAuth, nil, nil)
```
Comment (relyt0925) [inline control-plane-operator/controllers/hostedcontrolplane/pki/konnectivity.go:36]: this is a super small nit that you don't have to change and I may be wrong:

Should this just be `X509UsageClientAuth`? not clientServer

Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/pki/konnectivity.go:36]: and really be the same as the agent basically

**Comment by sjenning** [comment]
my testing this evening makes me think that `APIServices` are actually to working at the moment :confounded:

File: control-plane-operator/controllers/hostedcontrolplane/konnectivity/reconcile.go [hunk 2]
```diff
@@ -100,6 +100,8 @@ func buildKonnectivityServerContainer(image string) func(c *corev1.Container) {
 			strconv.Itoa(KonnectivityServerPort),
 			"--health-port=8092",
 			"--admin-port=8093",
+			"--mode=http-connect",
+			"--proxy-strategies=destHost,default",
```
Comment (relyt0925) [inline control-plane-operator/controllers/hostedcontrolplane/konnectivity/reconcile.go:104]: I think you want this to be defaultRoute

**Comment by relyt0925** [comment]
@sjennings I think maybe my one comment will help.....

File: control-plane-operator/controllers/hostedcontrolplane/konnectivity/reconcile.go (hunk 2 above)
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/konnectivity/reconcile.go:104]: And default actually should never be specified as it's not desired in any situation (default is send to any random backend)

File: control-plane-operator/controllers/hostedcontrolplane/konnectivity/reconcile.go [hunk 3]
```diff
@@ -281,6 +283,7 @@ func buildKonnectivityWorkerAgentContainer(image, host string, port int32) func(
 			host,
 			"--proxy-server-port",
 			fmt.Sprint(port),
+			"--agent-identifiers=default-route",
```
Comment (relyt0925) [inline control-plane-operator/controllers/hostedcontrolplane/konnectivity/reconcile.go:286]: I believe this should be default-route=true 

To properly enable

Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/konnectivity/reconcile.go:286]: example command line:
https://github.com/kubernetes-sigs/apiserver-network-proxy/commit/554a60b5bacec1e3c27f218d10ec004a244b1719#diff-06009d1be092147fe694efde678466f0c2f7771ba92e63ca27b912fe8e846606R78

File: control-plane-operator/controllers/hostedcontrolplane/konnectivity/reconcile.go (hunk 2 above)
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/konnectivity/reconcile.go:104]: defaultRoute deployment strategy:
https://github.com/kubernetes-sigs/apiserver-network-proxy/commit/554a60b5bacec1e3c27f218d10ec004a244b1719#diff-4ec4a95f03570d32223e630425a6daf105600a2bc4555dd38374307d09a67f30R47

(not default)

File: control-plane-operator/controllers/hostedcontrolplane/konnectivity/reconcile.go (hunk 3 above)
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/konnectivity/reconcile.go:286]: (and you don't specify default-route=true for the agents in the control plane if that makes sense)

**Comment by sjenning** [comment]
thanks @relyt0925 !

**Comment by sjenning** [comment]
@relyt0925 pushed some updates to explicitly list the ClusterIPs of the control plane APIServices in the `agent-identifiers` of the agent running in the control plane. However konnectivity still does not choose the correct backend.

```
$ oc logs konnectivity-agent-66b47c5657-sf4pr 
...
I0622 01:23:50.461287       1 options.go:93] AgentID set to 4e3ca10f-fa15-46e8-8754-8617a3c01943.
...
I0622 01:23:50.461327       1 options.go:98] AgentIdentifiers set to ipv4=172.30.57.224,ipv4=172.30.57.142,ipv4=172.30.53.34.
```
```
$ oc logs konnectivity-server-747b98f677-zl792 | grep -A3 -B3 4e3ca10f-fa15-46e8-8754-8617a3c01943
...
I0622 01:26:29.308877       1 server.go:617] "Connect request from agent" agentID="4e3ca10f-fa15-46e8-8754-8617a3c01943"
I0622 01:26:29.308920       1 backend_manager.go:191] "Register backend for agent" connection=&{ServerStream:0xc000018180} agentID="172.30.57.224,ipv4=172.30.57.142,ipv4=172.30.53.34"
...
```
Yet default route backend is used for IPs listed in the control plane agent
```
I0622 01:41:58.095377       1 tunnel.go:37] "Received request for host" method="CONNECT" host="172.30.57.224:443" userAgent=""
I0622 01:41:58.095403       1 tunnel.go:39] "TLS" commonName="konnectivity-client"
I0622 01:41:58.095470       1 tunnel.go:71] Set pending(rand=1752742903139208989) to &{0xc0003437c0 0xc000a63c00 {} 0x5103e0 true false false false <nil> {0xc000a8a620 map[] true true} map[] false 0 -1 200 false false [] 0 [84 117 101 44 32 50 50 32 74 117 110 32 50 48 50 49 32 48 49 58 52 49 58 53 56 32 71 77 84] [0 0 0 0 0 0 0 0 0 0] [50 48 48] 0xc00007d880 0}
I0622 01:41:58.095502       1 default_route_backend_manager.go:49] "Picked agent as backend" agentID="96471cab-e802-4eee-85b1-8fc7cdeb6988"
```

**Comment by sjenning** [comment]
It was pointed out that the delimiter for the `agent-identifiers` is `&`, not `,` :man_facepalming: 

Everything works now.

/test e2e-aws

**Comment by ironcladlou** [comment]
Awesome work, thanks!

/lgtm

**Pull request merged**